### PR TITLE
pass data and config to loss function

### PR DESCRIPTION
**Pull request by flomerz**
## Summary
Me again :) In this PR i want to add some small changes mainly for the hyperopt loss function. I did some more extended loss function. Which does some kind of cross validation over the the hole timeframe, to prevent overfitting.
So what i did for now is, i read it in by myself:
`Path("user_data") / 'hyperopt_results' / 'hyperopt_tickerdata.pkl'`
But I think that is an ugly workaround :(
So I think I just needed to pass the processed timeframe to the loss function to, to do that.
The same applies for backtesting stats. I want to analyze some splits of the timeframe and combine that to a more meaningfull score.

## Quick changelog

- pass processed timeframe to hyperopt loss function
- pass processed timeframe to backtest result stats

### CODE REVIEW

File: freqtrade/optimize/optimize_reports.py [hunk 1]
```diff
@@ -225,7 +225,7 @@ def generate_daily_stats(results: DataFrame) -> Dict[str, Any]:
 
 def generate_backtest_stats(btdata: Dict[str, DataFrame],
                             all_results: Dict[str, Dict[str, Union[DataFrame, Dict]]],
-                            min_date: Arrow, max_date: Arrow
+                            min_date: Arrow, max_date: Arrow, processed: Dict
```
Comment (xmatthias) [inline freqtrade/optimize/optimize_reports.py]: This will not make sense.

The (non-analyzed) backtest data is available as btdata should the raw candles be necessary - and keeping this data around will cause memory leaks when used with multiple strategies (`--strategy-list`).

File: freqtrade/optimize/backtesting.py [hunk 2]
```diff
@@ -445,11 +445,13 @@ def start(self) -> None:
 
         min_date = None
         max_date = None
+        processed = None
         for strat in self.strategylist:
-            min_date, max_date = self.backtest_one_strategy(strat, data, timerange)
+            min_date, max_date, processed = self.backtest_one_strategy(strat, data, timerange)
```
Comment (xmatthias) [inline freqtrade/optimize/backtesting.py]: Same as below, causing memory leaks (it's not fully leaking as we know it's still around) - but it'll increase the amount of memory necessary when using `--strategy-list` by at least 1x per strategy.

Since the data is not used anyway, there's no reason to do this.

File: freqtrade/optimize/hyperopt.py [hunk 3]
```diff
@@ -546,10 +546,11 @@ def generate_optimizer(self, raw_params: List[Any], iteration=None) -> Dict:
 
         )
         return self._get_results_dict(backtesting_results, min_date, max_date,
-                                      params_dict, params_details)
+                                      params_dict, params_details,
+                                      processed=processed)
```
Comment (xmatthias) [inline freqtrade/optimize/hyperopt.py:550]: I'm not sure what you'd need this for - and i have quite heavy fears that this will increase the amount of memory needed for hyperopt (hyperopt already has quite heavy memory requirements) - and if the need for memory increases even more it'll become unusable.

**Review by xmatthias — CHANGES_REQUESTED** [review]
Clearly the backtesting part is a no-go due to the memory-problems this will introduce.

In general, i can only recommend to open issues for topics as such, so the changes you would like to make can be discussed upfront, which will help to avoid wasting time on changes that don't make sense.

File: freqtrade/optimize/optimize_reports.py (hunk 1 above)
Comment (flomerz) [reply freqtrade/optimize/optimize_reports.py]: Agree :+1:  I don't know way I didn't saw this. Maybe because, I was on an older code-base while implementing it. I will remove it.

File: freqtrade/optimize/backtesting.py (hunk 2 above)
Comment (flomerz) [reply freqtrade/optimize/backtesting.py]: Not needed to, because of btdata from above.

File: freqtrade/optimize/hyperopt.py (hunk 3 above)
Comment (flomerz) [reply freqtrade/optimize/hyperopt.py:550]: I don't see why this should create a memory problem, because the data is already there. 
The time it gets free is when the function 'generate_optimizer' is finish. So it is just passed to the loss function before that. So there is no difference in memory usage.
I really need the data in the loss function or is there an other approach?

Second thing what i would need is the config in the loss function. Specially for the fee configuration. Is there a way to access it?

Comment (xmatthias) [reply freqtrade/optimize/hyperopt.py:550]: Fee is not part of the configuration (unless you use --fee), but is received from the exchange during startup of hyperopt/backtesting (check the `__init__()` method in backtesting).

for the memory part, i'll have to have a closer look, for the loss function it might be ok.

Comment (flomerz) [reply freqtrade/optimize/hyperopt.py:550]: Thanks for your effort :+1: 

I need the fee, because I do not use staking. Instead I have an all-in/all-out strategy.
So I need to calculate the fees based on the cumulative return.

Comment (xmatthias) [reply freqtrade/optimize/hyperopt.py:550]: Backtesting (nor hyperopt) does not support compounding at the moment - so your stake-amounts will always be the same, so i'm not quite sure what you'll calculate here...

I'm working on changing this (and it's almost complete - i'm just struggling with a few final calculations i need to adjust to have the result consistent).
so i'm not quite sure why the fee would be relevant (it'll be the same number for each exchange anyway ...)

Comment (flomerz) [reply freqtrade/optimize/hyperopt.py:550]: I look up the open/close times of the trades and search these in the btdata. Then I calculate the pct change over all periods and add the fee as penalty to the specific points. After that i calculate the product over all periods and then i get the cumulative return considering all fees.

What do you think about pass the btdata and the config object to the loss function?

Comment (xmatthias) [reply freqtrade/optimize/hyperopt.py:550]: as said, i think you're misjudging what config holds about fees.
fee is NOT in the configuration (at least not the fee that's applied).

It is however available as `fee_open` and `fee_close` in the trade results ... so i don't see why this would need config ... ?

Comment (flomerz) [reply freqtrade/optimize/hyperopt.py:550]: Sorry for confusion. I don't talk about the config.json. I talk about the self.config object.
Like here: https://github.com/freqtrade/freqtrade/blob/develop/freqtrade/optimize/backtesting.py#L105

Comment (xmatthias) [reply freqtrade/optimize/hyperopt.py:550]: i am aware of this, but the "fee" key is only read - not written to the configuration.
In the default case, where you don't use `--fee 0.0011` to override the exchange default fee - `config['fee']` will be empty, and `self.fee` will have the right value.

https://github.com/freqtrade/freqtrade/blob/develop/freqtrade/optimize/backtesting.py#L107

this `self.fee` object is then also assigned to each trade as `open_fee` and `close_fee`, and used/included in profit calculations.

So if you pass in a explicitly high fee, it'll be applied to all profit calculations by freqtrade - so by applying that again in your loss function, you're removing fees twice.

Comment (flomerz) [reply freqtrade/optimize/hyperopt.py:550]: Yes, but i don't use the profit calculation at all. Because it does not match to my use case. I set a relative high fee (containing fee + expected slippage) on each run with the cli argument and then i calculate the profit from scratch with the original time series. To consider compounding. So all i need for this is the btdata and the config accessible in the loss function. The fee is not the only config which I want to read in the loss function. I have some loss-function hyperparameters, which i would like to define in the config too. For now that are constants, but that is not ideal for different configurations.

**Review by xmatthias — APPROVED** [review]
Doesn't seem to impact performance after all

LGTM :+1: 

I've added the new parameter to the docs, otherwise noone except for you will be able to use it (without reading the code for "hidden" parameters)

**Comment by flomerz** [comment]
Thanks for your work, I appreciate it :+1: 
Docs is a good idea :)

Have you noticed to additional config parameter? I'm not sure if the sample strategies works this way..
